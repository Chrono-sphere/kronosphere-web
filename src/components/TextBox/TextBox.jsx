import React, { Component } from 'react';
import PropTypes from 'prop-types';
import uuidv4 from 'uuid/v4';
import $ from 'jquery';
import './TextBox.scss';

class TextBox extends Component {
  state = {
    value: '',
  };

  onInputChange = ({ target: { value } }) => {
    this.props.onValueChange(value);
    this.setState({ value }, () => this.isValid(this.state.value));
  };

  controlId = this.props.id ? this.props.id : uuidv4();

  isValid = (value) => {
    const badValidations = this.props.validations.filter(validation => !validation.isValid(value));
    const isValid = badValidations.length === 0;

    this.props.onValidityChange(isValid);

    if (this.props.showValidations && !isValid) {
      $(`#${this.controlId}`).css('border-right', '5px solid #f16b6b');
    } else {
      $(`#${this.controlId}`).css('border-right', '');
    }

    return isValid;
  }

  render = () => {
    const { value } = this.state;
    const { placeholder, inputType } = this.props;

    return (
      <div id={this.controlId} className="textbox">
        <input
          onChange={this.onInputChange}
          placeholder={placeholder}
          value={value}
          spellCheck={false}
          type={inputType}
        />
        <span className="input-highlight">
          {value.replace(/ /g, '\u00a0')}
        </span>
      </div>
    );
  }
}

TextBox.propTypes = {
  id: PropTypes.string,
  placeholder: PropTypes.string,
  validations: PropTypes.arrayOf(PropTypes.object),
  showValidations: PropTypes.bool,
  inputType: PropTypes.string,
  onValueChange: PropTypes.func,
  onValidityChange: PropTypes.func,
};

TextBox.defaultProps = {
  id: null,
  placeholder: '',
  validations: [],
  showValidations: true,
  inputType: 'normal',
  onValueChange: () => null,
  onValidityChange: () => null,
};

export default TextBox;
