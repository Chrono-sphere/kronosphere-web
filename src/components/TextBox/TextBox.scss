@import '~styles/colors';

.textbox {
    $input-font-size: 30px;
    $input-line-height: 70px;
    $font-family: Roboto Slab, sans-serif;

    position: relative;
    width: 500px;
    margin: 50px auto;

    input {
        height: 60px;
        width: 100%;
        min-width: 100%;
        padding: 0;
        border-radius: 0;
        line-height: $input-line-height;
        background-color: transparent;
        color: $colors-darkslategray;
        font-size: $input-font-size;
        border: none;
        outline: none;
        border-bottom: 3px solid $colors-darkslategray;
        font-family: $font-family;

        ::placeholder {
            color: $colors-darkslategray;
        }

        &:focus {
            + .input-highlight {
                border-top: 3px solid $colors-dodgerblue;
            }
        }
    }

    .input-highlight {
        font-size: $input-font-size;
        user-select: none;
        line-height: $input-line-height;
        position: absolute;
        left: 0;
        bottom: 0;
        max-width: 100%;
        height: 0;
        color: transparent;
        font-family: $font-family;
        overflow: hidden;
    }
}