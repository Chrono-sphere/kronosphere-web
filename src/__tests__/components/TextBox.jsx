import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import TextBox from 'components/TextBox/TextBox';

describe('TextBox', () => {
  const placeholder = 'text';

  it('should render without throwing an error', () => {
    expect(shallow(<TextBox placeholder={placeholder} />).find('div.textbox').exists()).toBe(true);
  });

  it('should set the props correctly', () => {
    const wrapper = shallow(<TextBox placeholder={placeholder} />);
    const { className, children } = wrapper.props();
    expect(className).toBe('textbox');
    expect(children).toBe(placeholder);
  });

  it('should render deep correctly', () => {
    const tree = renderer.create(<TextBox placeholder={placeholder} />);
    expect(tree).toMatchSnapshot();
  });
});
