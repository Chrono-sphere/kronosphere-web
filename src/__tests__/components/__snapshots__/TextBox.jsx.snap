// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextBox should render deep correctly 1`] = `
<div
  className="textbox"
  id="test"
>
  <input
    onChange={[Function]}
    placeholder="text"
    spellCheck={false}
    type="normal"
    value=""
  />
  <span
    className="input-highlight"
  >
    
  </span>
</div>
`;
